### ISSUE TITLE
variation of CLEVER score

### ISSUE THREAD

**Issue by kumarvrsec** [Trusted-AI/adversarial-robustness-toolbox#478]
When i tried to find CLEVER score of a classifier,I get different results when i run it again.

**Comment by beat-buesser** [comment]
Hi @kumarvrsec Thank you very much for using ART! Have you noticed that CLEVER contains a random sampling step? Could you please provide a minimal code example that reproduces your observation?

**Comment by kumarvrsec** [comment]
import keras
from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
import numpy as np
import tensorflow as tf
import torch.nn as nn
import torch.nn.functional as f
import torch.optim as optim

from art.estimators.classification.keras import KerasClassifier
from art.estimators.classification.pytorch import PyTorchClassifier
from art.estimators.classification.tensorflow import TensorFlowClassifier
from art.metrics.metrics import empirical_robustness, clever_t, clever_u, clever, loss_sensitivity, wasserstein_distance
from art.utils import load_mnist

from tests.utils import master_seed
batch_size, nb_train, nb_test = 100, 1000, 10
(x_train, y_train), (x_test, y_test), _, _ = load_mnist()
x_train, y_train = x_train[:nb_train], y_train[:nb_train]
x_test, y_test = x_test[:nb_test], y_test[:nb_test]
model = Sequential()
model.add(Conv2D(4, kernel_size=(5, 5), activation="relu", input_shape=(28, 28, 1)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(10, activation="softmax"))

model.compile(
    loss=keras.losses.categorical_crossentropy, optimizer=keras.optimizers.Adam(lr=0.01), metrics=["accuracy"]
)

# Get the classifier
krc = KerasClassifier(model=model, clip_values=(0, 1), use_logits=False)
# Get the classifier
krc.fit(x_train, y_train, batch_size=batch_size, nb_epochs=1, verbose=0)
R_L1 = 40
R_L2 = 2
# Test targeted clever
res0 = clever_t(krc, x_test[-1], 2, 10, 5, R_L1, norm=1, pool_factor=3)
print("Targeted CLEVER score:",res0)
res1 = clever_u(krc, x_test[-1], 10, 5, R_L1, norm=1, pool_factor=3)
print("Untargeted CLEVER score:",res1)
CLEVER untargeted:  22%|██▏       | 2/9 [00:00<00:00, 17.82it/s]Targeted CLEVER score: 8.250436166988266
CLEVER untargeted: 100%|██████████| 9/9 [00:00<00:00, 17.87it/s]Untargeted CLEVER score: 2.575134353783453

when i ran it again
the scores are 
CLEVER untargeted:  22%|██▏       | 2/9 [00:00<00:00, 13.30it/s]Targeted CLEVER score: 6.213340491310249
CLEVER untargeted: 100%|██████████| 9/9 [00:00<00:00, 11.85it/s]Untargeted CLEVER score: 1.5256504801629898

**Comment by kumarvrsec** [comment]
and also are there any generic hyperparameters for CLEVER

**Comment by beat-buesser** [comment]
When you run CLEVER for the second time, are you fitting the model again with a different random initialisation?

**Comment by kumarvrsec** [comment]
> When you run CLEVER for the second time, are you fitting the model again with a different random initialisation?

No,I am just running the same program again

**Comment by beat-buesser** [comment]
I think the program in the code example above is training a new model with different initialisation in every run in section "Get the classifier". This might be one explanation for the variations. I would recommend loading a pretrained model.

Another explanation is the random sampling inherent to CLEVER, try if you can reduce the variation with increasing `pool_factor` or setting seeds for Numpy.

**Comment by kumarvrsec** [comment]
Thanks,how to decide on the parameters given to clever model like radius etc

**Comment by Zephyrasgit** [comment]
Hey @kumarvrsec, me and my colleagues have tried to generate CLEVER scores for our model and dataset, and also have the same questions as you have. We were not sure what parameters to use, and why the scores are always different. But after talking to several people here on ART, and as well as the original authors of the paper, we have some findings that was quite helpful for us, so I thought I'd share them with you:

1. On scores being different after repeating the same code
For this, it is like what @beat-buesser has mentioned, there is a random sampling section in CLEVER, which is the part where the algorithm tries to sample datapoints using a reverse weibull distribution. I'm think this results in a non-deterministic generation of scores, i.e. scores tend to change, depending on how the sampling was initiated. To curb this, like what @beat-buesser mentioned, you can increase the `pool_factor`, or set numpy seeds.

2. On generic hyperparameters
This has been a question we've been trying to tackle for a long time. After talking to the paper's authors, they told us that it is best that you run several experiments to know which parameters is best suited for your model and dataset. Here is what we know: 
- We ran `clever_u` with increasing `radius`, while fixing `nb_batches` and `batch_size`. For us, we observe that the smaller the `radius`, the higher chance that the scores converge to the `radius`, since the estimated `CLEVER score is higher than the `radius`. As `radius` increases, the chances of scores converging to `radius` is less, and we pick the smallest `radius` such that scores do not converge to said `radius`. We also note that if your image is large, the `radius` should also be large.
- As for `nb_batches` and `batch_size`, we observe that it is a trade-off between accuracy of the scores, and time taken to generate the scores (our model is large, and our image size is large, 240X320X3, so algorithm takes longer when `nb_batches` and `batch_size` is large). We ran several generations of the scores, and we take the smallest `nb_batches` and `batch_size`, when the scores do not differ much between changes of `nb_batches` and `batch_size` respectively.
